src/mlpack/core/metrics/non_maximal_supression_impl.hpp

```diff
@@ -0,0 +1,126 @@
+
+  // Clear selected bounding boxes.
+  selectedBoundingBoxes.clear();
+
+  // Obtain Sorted indices for bounding boxes according to
+  // their confidence scores.
+  arma::ucolvec sortedIndices = arma::sort_index(confidenceScores);
+
+  // Pre-Compute area of each bounding box.
+  arma::mat area;
+  if (UseCoordinates)
+  {
+    area = (boundingBoxes.row(2) - boundingBoxes.row(0)) %
+        (boundingBoxes.row(3) - boundingBoxes.row(1));
+  }
+  else
+  {
+    area = (boundingBoxes.row(2)) % (boundingBoxes.row(3));
+  }
+
+  while (sortedIndices.n_elem > 0)
+  {
+    size_t selectedIndex = sortedIndices(sortedIndices.n_elem - 1);
+
+    // Choose the box with the largest probability.
+    selectedBoundingBoxes.insert_cols(0, boundingBoxes.col(selectedIndex));
+
+    // Check if there are other bounding boxes to compare with.
+    if (sortedIndices.n_elem == 1)
+    {
+      break;
+    }
+
+    // Remove the last index.
+    sortedIndices = sortedIndices(arma::span(0, sortedIndices.n_rows - 2),
+        arma::span());
+
+    // Calculate IoU of remaining boxes with the last bounding box with
+    // the highest confidence score.
+    BoundingBoxesType intersectionArea;
+    if (UseCoordinates)
+    {
+      intersectionArea = arma::clamp(arma::clamp(
+          boundingBoxes.submat(arma::uvec(1).fill(2), sortedIndices), DBL_MIN,
+          boundingBoxes(2, selectedIndex)) - arma::clamp(
+          boundingBoxes.submat(arma::uvec(1).fill(0), sortedIndices),
+          boundingBoxes(0, selectedIndex), DBL_MAX), 0.0, DBL_MAX) %
+          arma::clamp(arma::clamp(boundingBoxes.submat(arma::uvec(1).fill(3),
+          sortedIndices), DBL_MIN, boundingBoxes(3, selectedIndex)) -
+          arma::clamp(boundingBoxes.submat(arma::uvec(1).fill(1),
+          sortedIndices), boundingBoxes(1, selectedIndex), DBL_MAX),
+          0.0, DBL_MAX);
+    }
+    else
+    {
+      intersectionArea = arma::clamp(arma::clamp(
+          boundingBoxes.submat(arma::uvec(1).fill(2), sortedIndices) +
+          boundingBoxes.submat(arma::uvec(1).fill(0), sortedIndices), DBL_MIN,
+          boundingBoxes(2, selectedIndex) + boundingBoxes(0, selectedIndex)) -
+          arma::clamp(boundingBoxes.submat(arma::uvec(1).fill(0),
+          sortedIndices), boundingBoxes(0, selectedIndex), DBL_MAX), 0.0,
+          DBL_MAX) % arma::clamp(arma::clamp(
+          boundingBoxes.submat(arma::uvec(1).fill(3), sortedIndices) +
+          boundingBoxes.submat(arma::uvec(1).fill(1),
+          sortedIndices), DBL_MIN, boundingBoxes(3, selectedIndex) +
+          boundingBoxes(1, selectedIndex)) -
+          arma::clamp(boundingBoxes.submat(arma::uvec(1).fill(1),
+          sortedIndices), boundingBoxes(1, selectedIndex), DBL_MAX),
+          0.0, DBL_MAX);
+    }
+
+    BoundingBoxesType calculateIoU = intersectionArea /
+        (area(sortedIndices).t() - intersectionArea + area(selectedIndex));
+
+    sortedIndices.shed_rows(arma::find(calculateIoU > threshold));
```

Comment:
Linux doesn't seem to recognize this function. I'll try to find out why once the armadillo site is up. Other than this I think the rest is complete. I have added tests to show various cases.
Regards.

Reply:
<img width="834" alt="image" src="https://user-images.githubusercontent.com/26578203/82367685-0afc3b00-99c9-11ea-9001-47a9c09a0d60.png">

According to the Armadillo website :point_up:, `shed_rows()` for a vector is supported since Armadillo v9.400. And Our Azure Pipelines use the version 8.400.

We can use `shed_row()` for a vector, and below code will work.
```
arma::uvec indices = arma::find(calculateIoU > threshold);
  for (size_t i = 0; i < indices.size(); i++)
  {
    sortedIndices.shed_row(indices(i));
  }
```

Reply:
Ahh, Instead of using for loops I am now using the following :
```
sortedIndices = sortedIndices(arma::find(calculateIoU <= threshold));
```
The tests pass. I'll push with rest of the changes in half an hour.